---
title: "Analyzing and reshaping quantitative data"
author:
- name: Fabricio Almeida-Silva
  affiliation: VIB-UGent Center for Plant Systems Biology
output:
  BiocStyle::pdf_document
vignette: |
  %\VignetteIndexEntry{Analyzing and reshaping quantitative data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    warning = FALSE, 
    message = FALSE,
    eval = FALSE
)
```

# Overview

Here, I will describe the code to:

1. Read and parse quantitative data from __salmon__ output files.
2. Perform dimensionality reduction with UMAP and t-SNE.

First of all, let's load required packages and data.

```{r load_data, eval = TRUE}
set.seed(123) # for reproducibility

# Load packages
library(here)
library(bears)
library(GenomicFeatures)
library(SummarizedExperiment)
library(SingleCellExperiment)
library(scater)
library(scran)
library(DESeq2)
library(tidyverse)
library(patchwork)
```

```{r}
# Load data
load(here("data", "ds.rda"))
final_metadata <- read.csv(
    here("products", "tables", "final_metadata_classified_atlas_v2.tsv"),
    header = TRUE, sep = "\t"
) |>
    dplyr::select(-c(Run, Experiment))
```

# From salmon `quant.sf` files to `SummarizedExperiment`

Here, we will obtain a `SummarizedExperiment` object containing gene-level
transcript abundances in TPM and bias-corrected counts. Counts will be obtained
using the "bias correction without an offset" method from the Bioconductor
package `tximport`.

To create the `SummarizedExperiment` object, we will need a 2-column
data frame of transcript-to-gene mapping. Let's create it.

```{r tx2gene}
# Create a data frame of transcript-to-gene mapping
tx <- Biostrings::readDNAStringSet(
    here("data", "gmax_transcriptome.fa.gz")
)
tx2gene <- data.frame(
    TXNAME = gsub(" \\|.*", "", names(tx)),
    GENEID = gsub(".*\\| ", "", names(tx))
)

save(
    tx2gene, compress = "xz",
    file = here("products", "result_files", "tx2gene.rda")
)
```

Now, we can get the `SummarizedExperiment` object.

```{r salmon2se}
# Get gene-level transcript abundance estimates from salmon
## "Bias correction without an offset" method
se_atlas_gene <- salmon2se(
    final_metadata,
    level = "gene",
    salmondir = ds$salmondir,
    tx2gene = tx2gene
)
assay(se_atlas_gene, "gene_counts") <- round(
    assay(se_atlas_gene, "gene_counts") 
)

## "Original counts and offset" method
files <- file.path(ds$salmondir, final_metadata$BioSample, "quant.sf")
se_atlas_gene_offset <- tximport::tximport(
    files, type = "salmon", tx2gene = tx2gene
)
colnames(se_atlas_gene_offset$abundance) <- final_metadata$BioSample
colnames(se_atlas_gene_offset$counts) <- final_metadata$BioSample
colnames(se_atlas_gene_offset$length) <- final_metadata$BioSample

# Get transcript-level transcript abundance estimates from salmon
se_atlas_transcript <- salmon2se(
    final_metadata,
    level = "transcript",
    salmondir = ds$salmondir
)

# Save gene-level and transcript-level
save(
    se_atlas_gene, compress = "xz",
    file = here("products", "result_files", "se_atlas_gene.rda")
)

save(
    se_atlas_gene_offset, compress = "xz",
    file = here("products", "result_files", "se_atlas_gene_offset.rda")
)

save(
    se_atlas_transcript, compress = "xz",
    file = here("products", "result_files", "se_atlas_transcript.rda")
)

# Save final and complete sample metadata data frame
sample_metadata_complete <- as.data.frame(colData(se_atlas_gene))

save(
    sample_metadata_complete, compress = "xz",
    file = here("products", "result_files", "sample_metadata_complete.rda")
)
```

# Dimensionality reduction

Now, we will perform dimensionality reduction on highly variable genes 
with PCA, t-SNE, and UMAP. To increase speed and avoid noise, we will
use the PCs computed with the PCA as input to t-SNE and UMAP.

```{r load-counts}
# Load SummarizedExperiment object containing counts per gene
load(here("products", "result_files", "se_atlas_gene.rda"))
```

## Feature selection

To maximize biological signal and reduce noise, we will only use highly
variable genes for dimensionality reduction. Here, we will pick the top 5000
of genes with the highest biological components.

```{r feature-selection}
# Create a SingleCellExperiment with counts and log-normalized counts
atlas_counts_sce <- SingleCellExperiment(
    assays = list(
        counts = assay(se_atlas_gene, "gene_counts"),
        logcounts = log2(assay(se_atlas_gene, "gene_counts") + 1)
    ), 
    colData = colData(se_atlas_gene)
)

# Modeling the mean-variance relationship and visualizing the fit
mean_var_model <- modelGeneVar(atlas_counts_sce)
fit_mean_var <- metadata(mean_var_model)

p_fit_mean_var <- data.frame(
    mean = fit_mean_var$mean,
    var = fit_mean_var$var,
    trend = fit_mean_var$trend(fit_mean_var$mean)
) |>
    ggplot(aes(x = mean, y = var)) +
    geom_point(alpha = 0.4) +
    geom_line(aes(y = trend), color = "steelblue3", linewidth = 1.5) +
    labs(
        title = "Per-gene mean-variance relationship",
        subtitle = "Counts were normalized by library size and log-transformed",
        x = "Mean of log-expression", y = "Variance of log-expression"
    ) +
    theme_minimal()
```

```{r echo = FALSE}
# Save plot
save(
    p_fit_mean_var, compress = "xz",
    file = here("products", "plots", "p_fit_mean_var.rda")
)
```

```{r eval = TRUE}
# Load plot
load(here("products", "plots", "p_fit_mean_var.rda"))
```

```{r eval=TRUE}
p_fit_mean_var
```

```{r}
# Extract the top 5000 of genes with the highest biological components
hvg <- getTopHVGs(mean_var_model, n = 5000)
```

The object `hvg` is a character vector containing the IDs of the 
top 5000 genes with the highest biological components.  

## Principal components analysis (PCA)

Now, we will perform PCA using the genes in `hvg`.

```{r}
# Perform PCA
atlas_counts_sce <- fixedPCA(
    atlas_counts_sce, subset.row = hvg
)

# Plot proportion of variance explained by each PC
percent_var <- attr(reducedDim(atlas_counts_sce), "percentVar")

p_pca_percent_var <- data.frame(
    Variance = round(percent_var, 2),
    PC = factor(1:50, levels = 1:50)
) |>
    ggplot(aes(x = PC, y = Variance)) +
    geom_col(fill = "grey40") +
    geom_text(aes(label = Variance), hjust = -0.3) +
    labs(
        title = "Proportion of variance explained by each PC",
        x = "PC", y = "Variance explained (%)"
    ) +
    coord_flip() +
    theme_minimal() +
    ylim(0, 60)
```

```{r echo = FALSE}
# Save plot
save(
    p_pca_percent_var, compress = "xz",
    file = here("products", "plots", "p_pca_percent_var.rda")
)
```

```{r eval = TRUE, echo=FALSE}
# Load plot above
load(here("products", "plots", "p_pca_percent_var.rda"))
```

```{r eval = TRUE, fig.height=8, fig.width=6}
p_pca_percent_var
```

Based on the plot, we will use only the top 8 PCs for t-SNE and UMAP.

## *t*-stochastic neighbor embedding (t-SNE)

Now, we will perform dimensionality reduction with t-SNE using the top 8 PCs
obtained previously. We will first test running a t-SNE with 6 different
perplexity values: 10, 20, 30, 40, 50, 60. Then, we will select the best.

```{r}
# Get and plot t-SNE coordinates (perplexity = 10, 20, 30, 40, 50)
perplexities <- c(10, 20, 30, 40, 50, 60)
p_tsne <- lapply(perplexities, function(x) {
    
    tsne_coord <- runTSNE(
        atlas_counts_sce, perplexity = x,
        dimred = "PCA", n_dimred = 8
    )
    
    # Color by the variable "Part"
    p <- plotReducedDim(tsne_coord, dimred = "TSNE", colour_by = "Part") +
        labs(
            x = "t-SNE 1", y = "t-SNE 2",
            title = paste0("Perplexity = ", x)
        )
    
    return(p)
})

# Visualize all plots
p_tsne_all_perplexities_panel <- wrap_plots(p_tsne, nrow = 2) +
    plot_layout(guides = "collect") &
    ggsci::scale_color_d3("category20") &
    labs(color = "Part")
```

```{r echo = FALSE}
# Save plot
save(
    p_tsne_all_perplexities_panel, compress = "xz",
    file = here("products", "plots", "p_tsne_all_perplexities_panel.rda")
)
```

```{r eval = TRUE, echo = FALSE}
# Load plot
load(here("products", "plots", "p_tsne_all_perplexities_panel.rda"))
```

```{r eval=TRUE, fig.height=8, fig.width=12}
p_tsne_all_perplexities_panel
```

Based on the plots, we chose `perplexity = 60` as the best option. Now, let's
create an object containing only the plot for this perplexity value and
give it a better title.

```{r}
# Plot t-SNE with perplexity = 60
p_tsne_optimal_perplexity <- p_tsne_all_perplexities_panel[[6]] +
    labs(
        title = "t-SNE plot of all samples in the SEA 2.0",
        subtitle = "Coordinates were constructed from the top 8 principal components, with perplexity = 60"
    )
```

```{r echo = FALSE}
# Save plots
save(
    p_tsne_optimal_perplexity, compress = "xz",
    file = here("products", "plots", "p_tsne_optimal_perplexity.rda")
)
```

```{r eval=TRUE, echo=FALSE}
load(here("products", "plots", "p_tsne_optimal_perplexity.rda"))
```

```{r eval = TRUE}
p_tsne_optimal_perplexity
```


## Uniform manifold approximation and projection (UMAP)

Lastly, we will perform dimensionality reduction with UMAP using the
top 8 PCs identified before. Similarly to what we did for t-SNE, we will
run UMAP with 6 different values for the "number of neighbors" parameter:
10, 20, 30, 40, 50, and 60. Then, we will look at each plot to choose
the best.

```{r}
# Run UMAP with n_neighbors = 10, 20, 30, 40, 50
n_neighbors <- c(10, 20, 30, 40, 50, 60)
p_umap <- lapply(n_neighbors, function(x) {
    
    umap_coord <- runUMAP(
        atlas_counts_sce, n_neighbors = x,
        dimred = "PCA", n_dimred = 8
    )
    
    # Color by the variable "Part"
    p <- plotReducedDim(umap_coord, dimred = "UMAP", colour_by = "Part") +
        labs(
            x = "UMAP 1", y = "UMAP 2",
            title = paste0("Number of nearest neighbors = ", x)
        )
    
    return(p)
})

# Visualize all plots
p_umap_all_nneighbors_panel <- wrap_plots(p_umap, nrow = 2) +
    plot_layout(guides = "collect") &
    ggsci::scale_color_d3("category20") &
    labs(color = "Part")
```

```{r echo = FALSE}
# Save plot
save(
    p_umap_all_nneighbors_panel, compress = "xz",
    file = here("products", "plots", "p_umap_all_nneighbors_panel.rda")
)
```

```{r eval = TRUE, echo = FALSE}
# Load plot
load(here("products", "plots", "p_umap_all_nneighbors_panel.rda"))
```

```{r eval=TRUE, fig.height=8, fig.width=12}
p_umap_all_nneighbors_panel
```

Based on the plots, we chose `n_neighbors = 30` as the best option. Now, let's
create an object containing the final plot.

```{r}
# Plot UMAP with n_neighbors = 30
p_umap_optimal_nneighbors <- p_umap_all_nneighbors_panel[[3]] +
    labs(
        title = "UMAP plot of all samples in the SEA 2.0",
        subtitle = "Coordinates were constructed from the top 8 principal components, with n_neighbors = 30"
    )
```

```{r echo = FALSE}
# Save plots
save(
    p_umap_optimal_nneighbors, compress = "xz",
    file = here("products", "plots", "p_umap_optimal_nneighbors.rda")
)
```

```{r eval=TRUE, echo=FALSE}
load(here("products", "plots", "p_umap_optimal_nneighbors.rda"))
```

```{r eval = TRUE}
p_umap_optimal_nneighbors
```

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE, eval = TRUE}
sessioninfo::session_info()
```
